### PR TITLE
Feat/#41 Transformer 아키텍처에도 xAI 가능한 코드 구현 및  refactoring

### LINKED ISSUE #41: [feat] Transformer 아키텍처에도 xAI 가능한 코드 구현.
## 🔍 Issue Description
현재 GradCAM 은 CNN에서 기반으로 해서 Transformer 아키텍처에는 xAI 가 잘 안되고 있음. 그래서 Transformer 에 돌아가는 것...

## ✅ TO-DO
- [x]  Transformer 아키텍처에서 돌아가는 , xAI 찾고, 코드 구현.
- [x] 시각화 코드 def main과 분리, config.yaml 도 수정.
- [x] README 업데이트
- [x] 시각화 input 이미지 자동 결정 후, output 폴더에 자동 저장.

### PR DESCRIPTION
**Pull request by 2JAE22**
## 📍 Related Issue
- Resolved: #41 

## 👷 Work Description

**1. README.md 개선**
- 상세 설명 추가
- 자동 checkpoint_path 경로 제거
- 폴더 구조 가독성 향상

**2. 코드 구조 최적화**
- Gradcam 및 attention_mapping 함수를 src/utils/visual_utils.py로 이동
- main 함수에서 performance visualization만 유지하여 간결성 확보

**3. 설정 파일 업데이트**
- train/config에서 gradcam 관련 설정 제거
- train/config_CNN에 gradcam 설정 추가

**4. 시각화 기능 개선**
- viT attention 시각화 기능 구현.
- gradcam 및 vit 시각화 시 이미지 경로 자동 랜덤 선택 기능 구현
- 프로젝트 구조 정리

**5. settings 폴더 제거**
- License 파일을 프로젝트 루트 디렉토리로 이동

## 📣 To Reviewers

더 깔끔하게 고쳐주세요!

## ✅ Checklists
- [x] 필요없는 주석, 프린트문 제거했는지 확인
- [x] 컨벤션 지켰는지 확인

### CODE REVIEW

**Review by kim-minsol — APPROVED** [review]
리펙터링 수고하셨습니다!!!!! 우선 merge해주시고 추가 수정사항 있으면 제가 따로 이슈 팔게요!!

File: README.md [hunk 1]
```diff
@@ -59,17 +59,25 @@ terminal에서 `wandb login` 치고 나서 본인의 API 입력하면 됨.
...
-python train.py --config configs/train_configs/train/config.yaml
+python train.py --config configs/train_configs/train/config.yaml #Eva 모델
+python train.py --config configs/train_configs/train/config_CNN.yaml #Covnext 모델
 ```
-- **Wandb 쓰는 방법**
+- **Wandb 쓰는 방법** 👉 `--use_wandb`를 붙여쓰면 됩니다
 ```bash
-python train.py --config configs/train_configs/train/config.yaml --use_wandb
+python train.py --config configs/train_configs/train/config.yaml --use_wandb #test Eva모델
+python test.py --config configs/train_configs/test/config_CNN.yaml --use_wandb #test Covnext 모델
```
Comment (Puer-Hyun) [inline README.md:68]: test.py 코드가 github 상에는 보이질 않아 부득이하게 여기에 남기자면, test.py를 진행했을 때, accuracy, precision, recall 기타 등등을 logging하고, 해당 결과가 몇 이상일 때 자동으로 predict를 하여 submission.csv를 만드는 식으로 진행할 수 있습니다.

File: src/data/custom_datamodules/sketch_datamodule.py [hunk 2]
```diff
@@ -32,15 +32,19 @@ def setup(self, stage: Optional[str] = None):
             train_transforms = A.Compose(
                 [
                     A.Resize(height=448, width=448),
-                    #A.ShiftScaleRotate(shift_limit=(-25,25), scale_limit=0, rotate_limit=0, border_mode=3, p=0.5),
```
Comment (Puer-Hyun) [inline src/data/custom_datamodules/sketch_datamodule.py:35]: class SketchDataModule(BaseDataModule):
    def __init__(self, data_config_path: str, augmentation_config_path: str, seed: int):
        self.data_config = load_yaml_config(data_config_path)
        self.augmentation_config = load_yaml_config(augmentation_config_path)
        self.seed = seed  # TODO
        super().__init__(self.data_config)

    def setup(self, stage: Optional[str] = None):
        # 시드 설정
        torch.manual_seed(self.seed)

        if self.augmentation_config["augmentation"]["use_augmentation"]:
            train_transforms = self._get_augmentation_transforms()
        else:
            train_transforms = A.Compose(
                [
                    A.Resize(height=448, width=448),
                    A.ShiftScaleRotate(shift_limit=(-0.25, 0.25), scale_limit=0, rotate_limit=0, border_mode=3, p=0.5),
                    #A.ShiftScaleRotate(shift_limit=0, scale_limit=0, rotate_limit=(-30,30), border_mode=3, p=0.5),
                    A.GaussianBlur (blur_limit=(3,7), sigma_limit=30, p=0.5),
                    # A.RandomBrightnessContrast(brightness_limit=(-0.5, 0.05), contrast_limit=(-0.2, 0.1), p=0.5),
                    # A.VerticalFlip(p=0.3),
                    # A.HorizontalFlip(p=0.3),
                    #A.GridDropout(ratio=0.5, fill_value=255, p=0.5),
                    # A.RandomGridShuffle(grid=(3, 3), p=0.5),
                    A.Normalize(
                        mean=[0.485, 0.456, 0.406],
                        std=[0.229, 0.224, 0.225]),
                    ToTensorV2()

                ]
            )
            
            augmentation을 다룰 때도, 몇몇 augmentation을 삭제하고 추가하는 과정도 config에서 다룰 수 있습니다. 다음 프로젝트에서는 해당 부분 구현을 목표로 해보셔도 괜찮을 것 같습니다~!

File: src/data/custom_datamodules/sketch_datamodule.py [hunk 3]
```diff
@@ -32,15 +32,19 @@ def setup(self, stage: Optional[str] = None):
...
                     #A.ShiftScaleRotate(shift_limit=0, scale_limit=0, rotate_limit=(-30,30), border_mode=3, p=0.5),
-                    #A.GaussianBlur (blur_limit=(3,7), sigma_limit=30, p=0.5),
+                    A.GaussianBlur (blur_limit=(3,7), sigma_limit=30, p=0.5),
+                    # A.RandomBrightnessContrast(brightness_limit=(-0.5, 0.05), contrast_limit=(-0.2, 0.1), p=0.5),
+                    # A.VerticalFlip(p=0.3),
+                    # A.HorizontalFlip(p=0.3),
                     #A.GridDropout(ratio=0.5, fill_value=255, p=0.5),
-                    A.RandomGridShuffle(grid=(3, 3), p=0.5),
+                    # A.RandomGridShuffle(grid=(3, 3), p=0.5),
                     A.Normalize(
```
Comment (Puer-Hyun) [inline src/data/custom_datamodules/sketch_datamodule.py:43]: normalization을 할 때 반드시 '우리 데이터'에서도 [0.485,0.456,0.406], std = [0.229, 0.224, 0.225])로 진행해야할까요? 해당 부분 고민해보시면 또 좋을 것 같습니다.

File: src/data/custom_datamodules/sketch_datamodule.py [hunk 4]
```diff
@@ -32,15 +32,19 @@ def setup(self, stage: Optional[str] = None):
             train_transforms = A.Compose(
                 [
                     A.Resize(height=448, width=448),
-                    #A.ShiftScaleRotate(shift_limit=(-25,25), scale_limit=0, rotate_limit=0, border_mode=3, p=0.5),
+                    A.ShiftScaleRotate(shift_limit=(-0.25, 0.25), scale_limit=0, rotate_limit=0, border_mode=3, p=0.5),
```
Comment (Puer-Hyun) [inline src/data/custom_datamodules/sketch_datamodule.py:35]: ShiftSCaleRotate를 할 때, A.GaussianBlur를 할 때, 해당 파라미터들도 HyperParameter Tuning 단계에서 진행해볼 수 있을 것 같습니다 :)

File: src/data/custom_datamodules/sketch_datamodule.py [hunk 5]
```diff
@@ -32,15 +32,19 @@ def setup(self, stage: Optional[str] = None):
             train_transforms = A.Compose(
                 [
                     A.Resize(height=448, width=448),
```
Comment (Puer-Hyun) [inline src/data/custom_datamodules/sketch_datamodule.py:34]: 때에 따라 우리의 모델들은 항상 448,448로 Resize를 빈번하게 진행할 수도 있는데, '원본 이미지 파일들'이 3200, 4800 이런식으로 이미지의 사이즈가 큰 경우에는 미리 모든 이미지 파일들을 Resizing을 해두면 대규모 데이터를 훈련시킬 때 데이터 로딩 시간을 단축시킬 수도 있을 것 같습니다.

**Review by Puer-Hyun — COMMENTED** [review]
고생 많으셨습니다 :) 
코드를 클론해서 살펴보니 git 관리를 정말 잘하신 것 같습니다.
저도 귀찮아서 자주 생략하긴 하지만 commit message의 body도 채워주면 reviewer 입장에서 그리고 팀원 입장에서 편하긴 한 것 같습니다.

또, 커밋메세지를 남길 때 [Feat] [Fix] [Update]등의 구분을 잘 남겨주신 것 같아 아주 감동(?)이었습니다. 제가 감동할 건 아니지만?ㅋㅋㅋㅋ

더불어, Gitmoji 라는 툴이 있는데, vscode에 이 확장프로그램을 깔면 커밋 메세지 앞에 이모지를 달아서, 꼭 필요한 메세지만 보도록 할 수도 있어서 귀엽고 좋더라구요. 다음 프로젝트 때는 한 번 활용해보시면 좋을 것 같습니다.

고생하셨습니다.

File: src/data/custom_datamodules/sketch_datamodule.py (hunk 3 above)
Comment (kim-minsol) [reply src/data/custom_datamodules/sketch_datamodule.py:43]: 대회 데이터셋에 대한 mean/std를 구하여 Normalize 적용했을 때, 오히려 성능이 떨어지는 현상이 발생하였습니다:( 
다른 좋은 방법을 더 고민해보겠습니다..!
